File: Repository/MatchRepository.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using DatabaseConnection;
using Domain;
using Microsoft.Extensions.Logging;

namespace projectC.Repository
{
    public class MatchRepository
    {
        private readonly SQLiteConnection connection;
        private static readonly ILogger<MatchRepository> logger = LoggerFactory.Create(builder =>
        {
            builder.AddConsole();
        }).CreateLogger<MatchRepository>();

        public MatchRepository()
        {
            connection = DBRepo.Connect();
        }

        public void Add(Match match)
        {
            const string sql = "INSERT INTO matches (matchId, matchName, nr_seats, ticketPrice) VALUES (@id, @name, @seats, @price)";
            try
            {
                using var cmd = new SQLiteCommand(sql, connection);
                cmd.Parameters.AddWithValue("@id", match.matchId);
                cmd.Parameters.AddWithValue("@name", match.matchName);
                cmd.Parameters.AddWithValue("@seats", match.nr_seats);
                cmd.Parameters.AddWithValue("@price", match.ticketPrice);
                int rows = cmd.ExecuteNonQuery();
                if (rows > 0) logger.LogInformation("Match added successfully");
                else logger.LogWarning("Match not added");
            }
            catch (Exception ex)
            {
                logger.LogError("Error adding match: {Message}", ex.Message);
            }
        }

        public void Remove(int matchId)
        {
            const string deleteTicketsSQL = "DELETE FROM tickets WHERE matchId_fk = @id";
            const string deleteMatchSQL = "DELETE FROM matches WHERE matchId = @id";
            try
            {
                using var cmd1 = new SQLiteCommand(deleteTicketsSQL, connection);
                cmd1.Parameters.AddWithValue("@id", matchId);
                cmd1.ExecuteNonQuery();

                using var cmd2 = new SQLiteCommand(deleteMatchSQL, connection);
                cmd2.Parameters.AddWithValue("@id", matchId);
                int rows = cmd2.ExecuteNonQuery();
                if (rows > 0) logger.LogInformation("Match and tickets deleted");
                else logger.LogWarning("Match not deleted");
            }
            catch (Exception ex)
            {
                logger.LogError("Error deleting match: {Message}", ex.Message);
            }
        }

        public Match GetById(int matchId)
        {
            const string sql = "SELECT * FROM matches WHERE matchId = @id";
            try
            {
                using var cmd = new SQLiteCommand(sql, connection);
                cmd.Parameters.AddWithValue("@id", matchId);
                using var reader = cmd.ExecuteReader();
                if (reader.Read())
                {
                    return new Match(
                        reader.GetInt32(0),
                        reader.GetString(1),
                        reader.GetInt32(2),
                        reader.GetDouble(3));
                }
            }
            catch (Exception ex)
            {
                logger.LogError("Error getting match: {Message}", ex.Message);
            }
            return null;
        }

        public void Update(int matchId, Match match)
        {
            const string sql = "UPDATE matches SET matchName = @name, nr_seats = @seats, ticketPrice = @price WHERE matchId = @id";
            try
            {
                using var cmd = new SQLiteCommand(sql, connection);
                cmd.Parameters.AddWithValue("@name", match.matchName);
                cmd.Parameters.AddWithValue("@seats", match.nr_seats);
                cmd.Parameters.AddWithValue("@price", match.ticketPrice);
                cmd.Parameters.AddWithValue("@id", matchId);
                int rows = cmd.ExecuteNonQuery();
                if (rows > 0) logger.LogInformation("Match updated");
                else logger.LogWarning("Match not updated");
            }
            catch (Exception ex)
            {
                logger.LogError("Error updating match: {Message}", ex.Message);
            }
        }

        public List<Match> GetAll()
        {
            var matches = new List<Match>();
            const string sql = "SELECT * FROM matches ORDER BY ticketPrice DESC";
            try
            {
                using var cmd = new SQLiteCommand(sql, connection);
                using var reader = cmd.ExecuteReader();
                while (reader.Read())
                {
                    matches.Add(new Match(
                        reader.GetInt32(0),
                        reader.GetString(1),
                        reader.GetInt32(2),
                        reader.GetDouble(3)));
                }
            }
            catch (Exception ex)
            {
                logger.LogError("Error retrieving matches: {Message}", ex.Message);
            }
            return matches;
        }

        public bool SellTicketBySeats(int matchId, int number)
        {
            const string selectSql = "SELECT nr_seats FROM matches WHERE matchId = @id";
            const string updateSql = "UPDATE matches SET nr_seats = nr_seats - @count WHERE matchId = @id";

            try
            {
                using var selectCmd = new SQLiteCommand(selectSql, connection);
                selectCmd.Parameters.AddWithValue("@id", matchId);
                using var reader = selectCmd.ExecuteReader();
                if (reader.Read())
                {
                    int available = reader.GetInt32(0);
                    if (available >= number)
                    {
                        using var updateCmd = new SQLiteCommand(updateSql, connection);
                        updateCmd.Parameters.AddWithValue("@count", number);
                        updateCmd.Parameters.AddWithValue("@id", matchId);
                        int affected = updateCmd.ExecuteNonQuery();
                        if (affected > 0)
                        {
                            logger.LogInformation("Sold {number} tickets for match {matchId}");
                            return true;
                        }
                    }
                    else
                    {
                        logger.LogWarning("Not enough seats for match {matchId}");
                        return false;
                    }
                }
            }
            catch (Exception ex)
            {
                logger.LogError("Error selling tickets: {Message}", ex.Message);
            }
            return false;
        }
    }
}

File: project-java-BogdanBala112/HomeAssignment1/src/main/java/Repository/MatchRepository.java
package Repository;

import Entities.Match;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class MatchRepository implements GenericInterface<Match, Integer>{
    private static final Logger logger = LogManager.getLogger(MatchRepository.class);

    public Connection connection;

    public MatchRepository() {
        this.connection = DBAccess.connect();
    }

    @Override
    public void add(Match match){
        String sql = "INSERT INTO matches (matchId, matchName, nr_seats, ticketPrice) VALUES (?, ?, ?, ?)";

        try (PreparedStatement stmt = connection.prepareStatement(sql)) {
            stmt.setInt(1, match.getMatchId());
            stmt.setString(2, match.getMatchName());
            stmt.setInt(3, match.getNr_seats());
            stmt.setDouble(4, match.getticketPrice());

            int rowsInserted = stmt.executeUpdate();
            if (rowsInserted > 0)
                logger.info("Match added successfully");
            else
                logger.warn("Match was not added");

        } catch (SQLException e) {
            logger.error("Error adding match: " + e.getMessage());
        }
    }

    @Override
    public void remove(Integer matchId){
        String sql = "DELETE FROM matches WHERE matchId = ?";
        String deleteTicketsSQL = "DELETE FROM tickets WHERE matchId_fk = ?";

        try(PreparedStatement stmt = connection.prepareStatement(sql) ;
        PreparedStatement stmt2 = connection.prepareStatement(deleteTicketsSQL)) {
            stmt2.setInt(1, matchId);
            stmt2.executeUpdate();

            stmt.setInt(1,matchId);
            int rowsDeleted = stmt.executeUpdate();
            if (rowsDeleted > 0)
                logger.info("Match and associated tickets deleted successfully.");
            else
                logger.warn("Match was not removed");
        }
        catch (SQLException e) {
            logger.error("Error deleting match: " + e.getMessage());
        }
    }

    @Override
    public Match getById(Integer matchId) {
        String sql = "SELECT * FROM matches WHERE matchId = ?";

        try(PreparedStatement stmt = connection.prepareStatement(sql)){
            stmt.setInt(1,matchId);
            ResultSet resultSet = stmt.executeQuery();

            if (resultSet.next()) {
                return new Match(
                        resultSet.getInt("matchId"),
                        resultSet.getString("matchName"),
                        resultSet.getInt("nr_seats"),
                        resultSet.getDouble("ticketPrice")
                );
            }
        }
        catch (SQLException e) {
            logger.error("Error retrieving the desired match: " + e.getMessage());
        }
        return null;
    }

    @Override
    public void update(Integer matchId, Match match){
        String sql = "UPDATE matches SET matchName = ?, nr_seats = ?, ticketPrice = ? WHERE matchId = ?";

        try(PreparedStatement stmt = connection.prepareStatement(sql)){
            stmt.setString(1,match.getMatchName());
            stmt.setInt(2,match.getNr_seats());
            stmt.setDouble(3,match.getticketPrice());
            stmt.setInt(4,match.getMatchId());

            int rowsUpdated = stmt.executeUpdate();

            if (rowsUpdated > 0)
                logger.info("Match updated successfully");
            else
                logger.warn("Match was not updated");
        }
        catch(SQLException e){
            logger.error("Error updating match: " + e.getMessage());
        }
    }

    @Override
    public List<Match> getAll(){
        List<Match> matches = new ArrayList<>();
        String sql = "SELECT * FROM matches ORDER BY ticketPrice DESC";

        try(PreparedStatement stmt = connection.prepareStatement(sql)){
                ResultSet resultSet = stmt.executeQuery();
                while (resultSet.next()){
                    Match match = new Match(
                            resultSet.getInt("matchId"),
                            resultSet.getString("matchName"),
                            resultSet.getInt("nr_seats"),
                            resultSet.getInt("ticketPrice")
                    );
                    matches.add(match);
                }
        }
        catch (SQLException e) {
            logger.error("Error retrieving all matches: " + e.getMessage());
        }
        return matches;
    }

    public boolean sellTicketBySeats(int matchId, int number){
        String selectSql = "SELECT nr_seats FROM matches WHERE matchId = ?";
        String updateSql = "UPDATE matches SET nr_seats = nr_seats - ? WHERE matchId = ?";

        try (PreparedStatement selectStmt = connection.prepareStatement(selectSql); PreparedStatement updateStmt = connection.prepareStatement(updateSql)){
            selectStmt.setInt(1, matchId);
            ResultSet rs = selectStmt.executeQuery();

            if (rs.next()){
                int availableSeats = rs.getInt("nr_seats");
                if (availableSeats >= number){
                    updateStmt.setInt(1, number);
                    updateStmt.setInt(2, matchId);
                    int rowsAffected = updateStmt.executeUpdate();

                    if (rowsAffected > 0){
                        logger.info("Sold: " + number  + " tickets for match " + matchId);
                        return true;
                    }
                }
                else {
                    logger.warn("Not enough tickets to sell!");
                    return false;
                }
            }

        }
        catch (SQLException e){
            logger.error("Error updating match: " + e.getMessage());
        }

        return false;

    }

}
